File: scenarios/tunnel-manager/api/domain/recovery.go
package domain

import "time"

// RecoveryState tracks the current state of the recovery engine.
type RecoveryState struct {
	Status         string    `json:"status"`          // "idle", "monitoring", "recovering", "circuit_open"
	ConsecFailures int       `json:"consec_failures"` // consecutive health check failures
	BackoffLevel   int       `json:"backoff_level"`   // current backoff level
	FailedRecovery int       `json:"failed_recoveries"`
	CircuitOpen    bool      `json:"circuit_open"`
	LastCheck      time.Time `json:"last_check"`
	LastRecovery   time.Time `json:"last_recovery,omitempty"`
	NextRetryAfter time.Time `json:"next_retry_after,omitempty"`
}

// RecoveryEvent records a recovery attempt.
type RecoveryEvent struct {
	ID          int       `json:"id"`
	TriggerType string    `json:"trigger_type"` // "ready_failure", "ha_connection_loss", "manual"
	Action      string    `json:"action"`       // "systemctl_restart"
	Outcome     string    `json:"outcome"`      // "success", "failure", "skipped"
	Details     string    `json:"details,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
}

File: scenarios/tunnel-manager/ui/src/test-utils/mocks/recovery.ts
/**
 * Mock builders for `api/recovery`. See `test-utils/mocks/api.ts` for the
 * builder/hoisting rationale.
 */
import { vi } from "vitest";
import { timestampFromDate } from "@bufbuild/protobuf/wkt";

export interface RecoveryMocks {
  recoveryClient: {
    getState: ReturnType<typeof vi.fn>;
    listEvents: ReturnType<typeof vi.fn>;
    recover: ReturnType<typeof vi.fn>;
  };
}

export const makeRecoveryState = (overrides: Record<string, unknown> = {}) => ({
  status: 2, // MONITORING
  consecFailures: 0,
  backoffLevel: 0,
  failedRecoveries: 0,
  circuitOpen: false,
  lastCheck: timestampFromDate(new Date(2026, 5, 18, 12, 0, 0)),
  lastRecovery: undefined,
  nextRetryAfter: undefined,
  ...overrides,
});

export const makeRecoveryEvent = (overrides: Record<string, unknown> = {}) => ({
  id: "event-1",
  trigger: "ready_failure",
  action: "systemctl_restart",
  outcome: 1, // SUCCESS
  details: "restarted cloudflared",
  attempt: 1,
  createdAt: timestampFromDate(new Date(2026, 5, 18, 11, 0, 0)),
  ...overrides,
});

export const makeRecoveryMocks = (): RecoveryMocks => ({
  recoveryClient: {
    getState: vi.fn().mockResolvedValue({ state: makeRecoveryState() }),
    listEvents: vi.fn().mockResolvedValue({ events: [] }),
    recover: vi.fn().mockResolvedValue({ outcome: 1, event: makeRecoveryEvent() }),
  },
});
